File: protobuf-serialization/src/test/java/com/stannis/benchmark/bench/RESTBenchmark.java
package com.stannis.benchmark.bench;

import com.stannis.benchmark.mapper.RecipeProtoMapper;
import com.stannis.benchmark.model.Ingredient;
import com.stannis.benchmark.model.IngredientUsed;
import com.stannis.benchmark.model.MeasurementType;
import com.stannis.benchmark.model.Recipe;
import com.stannis.benchmark.util.RecipeTestUtil;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import generated.stannis.benchmark.Messages;

import org.openjdk.jmh.annotations.*;
import org.springframework.http.*;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.http.converter.protobuf.ProtobufHttpMessageConverter;
import org.springframework.web.client.RestTemplate;

import java.util.List;
import java.util.concurrent.TimeUnit;

@State(Scope.Benchmark)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
public class RESTBenchmark {

    private static final String JSON_URL = "http://localhost:8080/json/add";
    private static final String PROTO_URL = "http://localhost:8080/proto/add";

    private RestTemplate template;
    private HttpHeaders jsonHeaders;
    private HttpHeaders protoHeaders;
    private Recipe recipe;
    private Messages.Recipe protoRecipe;

    public RESTBenchmark() {
        List<HttpMessageConverter<?>> messageConverters = Lists.newArrayList();
        messageConverters.add(new ProtobufHttpMessageConverter());
        messageConverters.add(new MappingJackson2HttpMessageConverter());

        template = new RestTemplate(messageConverters);

        jsonHeaders = new HttpHeaders();
        jsonHeaders.setContentType(MediaType.APPLICATION_JSON);

        protoHeaders = new HttpHeaders();
        protoHeaders.setContentType(ProtobufHttpMessageConverter.PROTOBUF);
    }

    @Setup
    public void setup() {
        IngredientUsed jalepenoUsed = new IngredientUsed(new Ingredient("Jalepeno", "Spicy Pepper"), MeasurementType.ITEM, 1);
        IngredientUsed cheeseUsed = new IngredientUsed(new Ingredient("Cheese", "Creamy Cheese"), MeasurementType.OUNCE, 4);

        recipe = RecipeTestUtil.createRecipe("My Recipe", "Some spicy recipe using a few items", ImmutableList.of(jalepenoUsed, cheeseUsed));
        protoRecipe = RecipeProtoMapper.parseAsProto(recipe);
    }

    @Benchmark
    public Integer store_receipt_via_json() {
        ResponseEntity<Integer> response = template.exchange(JSON_URL, HttpMethod.POST, new HttpEntity<>(recipe, jsonHeaders), Integer.class);

        return response.getBody();
    }

    @Benchmark
    public Integer store_receipt_via_proto() {
        ResponseEntity<Integer> response = template.exchange(PROTO_URL, HttpMethod.POST, new HttpEntity<>(protoRecipe, protoHeaders), Integer.class);

        return response.getBody();
    }

}

File: protobuf-bench/src/test/kotlin/com/stannis/protobufbench/benchmark/RESTBenchmark.kt
package com.stannis.protobufbench.benchmark

import com.google.common.collect.ImmutableList
import com.google.common.collect.Lists
import com.stannis.protobufbench.model.Ingredient
import com.stannis.protobufbench.model.IngredientUsed
import com.stannis.protobufbench.model.MeasurementType
import com.stannis.protobufbench.model.Recipe
import generated.stannis.proto.Messages
import org.openjdk.jmh.annotations.*
import org.springframework.http.*
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter
import org.springframework.http.converter.protobuf.ProtobufHttpMessageConverter
import org.springframework.web.client.RestTemplate
import java.util.concurrent.TimeUnit


@State(Scope.Benchmark)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
class RESTBenchmark {
    private val template: RestTemplate
    private val jsonHeaders: HttpHeaders
    private val protoHeaders: HttpHeaders
    private var recipe: Recipe? = null
    private var protoRecipe: Messages.Recipe? = null

    init {
        val messageConverters: MutableList<HttpMessageConverter<*>> = Lists.newArrayList()
        messageConverters.add(ProtobufHttpMessageConverter())
        messageConverters.add(MappingJackson2HttpMessageConverter())
        template = RestTemplate(messageConverters)
        jsonHeaders = HttpHeaders()
        jsonHeaders.contentType = MediaType.APPLICATION_JSON
        protoHeaders = HttpHeaders()
        protoHeaders.contentType = ProtobufHttpMessageConverter.PROTOBUF
    }

    @Setup
    fun setup() {
        val jalepenoUsed = IngredientUsed(Ingredient("Jalepeno", "Spicy Pepper"), MeasurementType.ITEM, 1)
        val cheeseUsed = IngredientUsed(Ingredient("Cheese", "Creamy Cheese"), MeasurementType.OUNCE, 4)
        recipe = RecipeTestUtil.createRecipe(
            "My Recipe",
            "Some spicy recipe using a few items",
            ImmutableList.of(jalepenoUsed, cheeseUsed)
        )
        protoRecipe = RecipeProtoMapper.parseAsProto(recipe)
    }

    @Benchmark
    fun store_receipt_via_json(): Int? {
        val response = template.exchange(
            JSON_URL, HttpMethod.POST, HttpEntity<Any>(recipe, jsonHeaders),
            Int::class.java
        )
        return response.body
    }

    @Benchmark
    fun store_receipt_via_proto(): Int? {
        val response = template.exchange(
            PROTO_URL, HttpMethod.POST, HttpEntity<Any>(protoRecipe, protoHeaders),
            Int::class.java
        )
        return response.body
    }

    companion object {
        private const val JSON_URL = "http://localhost:8080/json/add"
        private const val PROTO_URL = "http://localhost:8080/proto/add"
    }
}

